doc/functions/fileset.section.md

```diff
@@ -0,0 +1,130 @@
+unions [
+  (singleton ./Makefile)
+  (singleton ./src)
+]
+```
+
+For convenience, all fileset operations implicitly coerce path arguments to file sets using `lib.singleton`, so the above is the same as
+```nix
+with lib.fileset;
+unions [
+  ./Makefile
+  ./src
+]
+```
+
+Another way to create file sets is by transforming existing file sets, e.g. using `lib.fileset.suffixFileFilter`:
+```
+with lib.fileset;
+union [
+  ./Makefile
+  (suffixFileFilter ["c" "h"] ./src)
+]
+```
+
+A fileset can be used in a derivation build by passing it as a `src` to any `stdenv.mkDerivation`-based builder. It is then necessary to also set `sourceRoot` to the path of the builder entry point.
+```nix
+stdenv.mkDerivation {
+  src = lib.sources.fromFileset {
+    # By default "source"
+    name = "myProject";
+    entryPoint = ./.;
+    # By default, the root is set to the entry point
+    # Such that when files outside the entry point need to be included,
+    root = ../.;
+  } (l: with l;
+    unions [
+      ./.
+    ]
+  );
+
+  src =
+    with lib.fileset;
+    unions [
+      ./Makefile
+      (suffixFileFilter ["c" "h"] ./src)
+    ];
+  sourceRoot = ./.;
+  postUnpack = ''
+    cat ./Makefile
+    ls ./src
+  '';
+}
+```
+
+By setting `sourceRoot` to `./.`, the builder can access the files at paths relative to directory of the file that defines it.
+
+Note that only exactly the files in the file set will be imported into the store and influence the derivation hash.
+
+File sets are not restricted to just files in the current directory though, any file can be included. If you're in a project subdirectory this means that you can use also include files from the project root directory:
+```nix
+stdenv.mkDerivation {
+  src =
+    with lib.fileset;
+    unions [
+      ./Makefile
+      (suffixFileFilter ["c" "h"] ./src)
+      ../README.md
+    ];
+  sourceRoot = ./.;
+  postUnpack = ''
+    cat ./Makefile
+    ls ./src
+    cat ../README.md
+  '';
+}
+```
+
+Warning: Any intermediate directory names will influence the store hash.
+
+Here's a more extensive example showcasing some file set functions:
+
+```nix
+stdenv.mkDerivation {
+
+  src =
+    with lib.fileset;
+
+    # Intersection between two sets of files
+    intersect
+
+      # All files tracked by the git root in the parent
+      (gitTracked ../.)
+
+      # Remove all files in "manual" directories
+      (lib.flip difference (directoryFilter "manual" ./.)
+
+        # Union between a sets of files
+        (unions [
+
+          # A single file from the current and parent directory
+          ./Makefile
+          ../LICENSE
+
+          # All files in the ./tests directory
+          ./tests
+
+          # All files in the ./src directory with suffixes ".c" and ".h"
+          (fileExtensionFilter ["c" "h"] ./src)
+        ])
+      );
+
+  sourceRoot = ./.;
+}
+```
+
+To see everything you can do with file sets, check out the [reference documentation](#sec-functions-library-fileset).
```

Comment:
```suggestion
To see everything you can do with file sets, see the [library documentation](#sec-functions-library-fileset).
```

We're already in reference documentation; just an introductory section.

Reply:
I wrote this in a tutorial-style, with the expectation that it would be moved to (or just rendered at) nix.dev at some point (that's where the doc team is moving tutorial docs towards).

Reply:
While all text can be learned from, this is not a "learning experience", certainly not according to [diataxis](https://diataxis.fr/tutorials/). And I think that's fine. File sets aren't just functions; the type needs reference documentation too, and it's a good place for a basic introduction to the concept.

> In general, tutorials are the weakest part of documentation, the most misunderstood and the most difficult to do well. Most software projects have poor - or non-existent - tutorials.

Reply:
Well then I clearly don't know how to write a tutorial :sweat_smile:, please take another look at the document now though, I wouldn't really count that as part of the reference.